### DISCUSSION TITLE
Syncing fish history across computers

### DISCUSSION THREAD

**Discussion by ilyagr** [fish-shell/fish-shell#9420] [Q&A]
Let's say I have `fish_history` and `another_computer_history` files next to each other. What's the best way to import `another_computer_history` into `fish_history`?

Will `cat another_computer_history >> fish_history && history merge` work?

Would it be useful to have a `history merge --from another_computer_history` command?

Is there a better way?

Additional background: In my case, I usually make `fish_history` a symlink to `computer_name_history` on one computer a symlink to `another_computer_history` on another computer. I then keep these files for various computers in a git repository. Sometimes I want two of the computers to share history, but I don't want to lose anything.

Thanks!

**Comment by JanPokorny** [comment]
You could try https://atuin.sh/, if it fits your use-case -- it's sort of a battery-included solution for shell history sync

**Comment by digitalsignalperson** [comment]
I'm also interested in the multiple computers syncing, but I think this also generally applies to a single computer and is a missing feature related to `set -x fish_history "session_name"`. 

E.g. following the docs https://fishshell.com/docs/current/cmds/history.html#customizing-the-name-of-the-history-file if you use this feature and do `set -x fish_history "fun"` you will have both `$XDG_DATA_HOME/fish/fish_history` and `$XDG_DATA_HOME/fish/fun_history`. But I don't see any suggested way to merge these files.

I'd like to just sync the fish_history files from multiple computers in a similar way and have them side by side like `$XDG_DATA_HOME/fish/computer{1,2}_history`. Something like atuin.sh seems too heavy for simply merging two plain text files that are next to each other in a folder. 

I think the issue with `cat another_computer_history >> fish_history` is that fish will use flock() to hold access to the history file while it writes to it. That cat might get interrupted halfway and result in broken lines.  Maybe we only need a couple lines of a shell script that can flock the history file while we merge multiple files periodically.

history.cpp
```
        // Exclusive lock on the entire file. This is released when we close the file (below). This
        // may fail on (e.g.) lockless NFS. If so, proceed as if it did not fail; the risk is that
        // we may get interleaved history items, which is considered better than no history, or
        // forcing everything through the slow copy-move mode. We try to minimize this possibility
        // by writing with O_APPEND.
        maybe_lock_file(fd.fd(), LOCK_EX);
```

**Reply by digitalsignalperson** [reply]
Here's an attempt at a python script to do this. I haven't tested it yet, but I will once I have the file syncing set up how I want. It might work well with the other fish history mounted over sshfs so that histories can be locked on both computers to prevent broken/interleaved items.

https://gist.github.com/digitalsignalperson/e6828c55748d2ddf214ffa75c57ea5fa

**Reply by stephanedaviet** [reply]
What about doing the history merge in another shell? This way, Fish will not lock the history to write the concat command.

**Comment by pcarrier** [comment]
Made the tiny https://github.com/pcarrier/plenty
